crates/services/p2p/src/peer_manager.rs

```diff
@@ -453,3 +475,217 @@ impl NetworkBehaviour for PeerInfoBehaviour {
+            .filter(|peer_id| !self.connected_peers.contains_key(peer_id))
+    }
+
+    fn reserved_peers_connected_count(&self) -> usize {
+        self.reserved_peers.iter().fold(0, |count, peer_id| {
+            if self.connected_peers.contains_key(peer_id) {
+                count + 1
+            } else {
+                count
+            }
+        })
+    }
+
+    fn find_disconnected_reserved_peer(&self) -> Option<PeerId> {
+        self.reserved_peers
+            .iter()
+            .find(|peer_id| self.connected_peers.contains_key(peer_id))
+            .cloned()
+    }
+
+    /// Handles the first connnection established with a Peer
+    fn handle_initial_connection(&mut self, peer_id: PeerId) {
+        // if the connected Peer is not from the reserved peers
+        if !self.reserved_peers.contains(&peer_id) {
+            let non_reserved_peers_connected =
+                self.connected_peers.len() - self.reserved_peers_connected_count();
+
+            // check if there is no more space for non-resereved peers
+            if non_reserved_peers_connected >= self.non_reserved_peers_allowed {
+                // todo/potential improvement: once `Peer Reputation` is implemented we could check if there are peers
+                // with poor reputation and disconnect them instead?
+
+                // Too many peers already connected, disconnect the Peer
+                self.pending_events.push_back(PeerInfoEvent::TooManyPeers {
+                    peer_to_disconnect: peer_id,
+                    peer_to_connect: self.find_disconnected_reserved_peer(),
+                });
+
+                // early exit, we don't want to report new peer connection
+                // nor insert it in `connected_peers`
+                // since we're going to disconnect it anyways
+                return
+            }
+        }
+
+        // insert and report on new Peer Connection
+        // for either, reserved peer or non-reserved
+        self.connected_peers.insert(peer_id, PeerInfo::default());
+        self.pending_events
+            .push_back(PeerInfoEvent::PeerConnected(peer_id));
+    }
+
+    /// Handles on peer's last connection getting disconnected
+    fn handle_peer_disconnect(&mut self, peer_id: PeerId) {
+        // try immediate reconnect if it's a reserved peer
+        let should_reconnect = self.reserved_peers.contains(&peer_id);
+        self.connected_peers.remove(&peer_id);
+
+        self.pending_events
+            .push_back(PeerInfoEvent::PeerDisconnected {
+                peer_id,
+                should_reconnect,
+            })
+    }
+}
+
+fn log_missing_peer(peer_id: &PeerId) {
+    debug!(target: "fuel-libp2p", "Peer with PeerId: {:?} is not among the connected peers", peer_id)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn get_random_peers(size: usize) -> Vec<PeerId> {
+        (0..size).map(|_| PeerId::random()).collect()
+    }
+
+    #[test]
+    fn test_peer_manager_struct() {
```

Comment:
This test should probably be split into smaller units, it's not immediately obvious what being covered here or what the intention of the test is without reading through each assertion.

Reply:
separated into 3 different tests